M2/Macaulay2/packages/Permutations/Permutations.m2

```diff
@@ -0,0 +1,311 @@
+-- Sean Grate
+
+-----------------------------------------------------------------------------
+-- Local utilities
+-----------------------------------------------------------------------------
+
+to1Index = w -> (w / (i -> i+1))
+to0Index = w -> (w / (i -> i-1))
+
+-----------------------------------------------------------------------------
+-- Permutation type declarations and basic constructors
+-----------------------------------------------------------------------------
+
+Permutation = new Type of BasicList
+Permutation.synonym = "permutation"
+
+isValidPermutation = method(TypicalValue => Boolean)
+isValidPermutation List := Boolean => w -> (((sort w) == toList(1..#w)) and not (#w == 0))
+
+new Permutation from BasicList := (typeofPermutation,w) -> (
+    if not isValidPermutation(w) then error(toString w | " is not a valid permutation in one-line notation.");
+    w
+)
```

Comment:
A better approach is this: don't check for correctness in the constructor, because it will slow down people's code (e.g. if they're making lots of very large permutations, and they know their permutations are definitely valid, then there's no need to check); instead change your verification method to `isWellDefined Permutation`, so people can verify that the permutation they constructed is correct.

Also, I think you might want `VisibleList` instead of `BasicList` in both places?

Reply:
That makes sense and should be corrected in some new commits. Related, but can I also use `VisibleList` as a type anywhere I would expect a `List` and `Sequence` to behave identically? I would hope that this would mean some code could be condensed and reduce the total number of overloads for a method.

Reply:
Yes, List, Sequence, Array, and AngleBarList inherit from VisibleList. On the other hands, loads of things inherit from BasicList (think Option, Command, URL, etc.)